### PR TITLE
Advanced Payments API

### LINKED ISSUE #1956: Advanced Payments
## Overview

The Advanced Payments feature will offer a comprehensive suite of payment functionalities catering to different payment types and scenarios, including a payment builder for completely flexibility, batch payments for bulk import and payments at once, split payments for distributing funds, staged payments for milestones, and streaming payments for continuous payments. This will empower users to efficiently manage fund allocation and distribution.

## User stories

- As a USER, managing payments, I would like to be able to create multiple payments at once to multiple people and in multiple token types.
- As a USER, managing payments, I would like to be able to have different ways of
making payments, such as splitting funds between recipients, doing batch upload
payments, making payments based on milestones, and making payments continually (e.g. a salary).
- As a USER, I would like to have advanced payments go through multiple checks to ensure legitimacy and so that no mistakes are made.
- As a USER, I would like to be able to see the details of the payment, who has participated and what stage it is in.
- As a USER, I would like to be able to include a details such as a title and description to explain the purpose of the payment.
- As a USER, I would like to be able to see the types of payments being made and all the details of the payment including recipients, values and reasoning for a payment.
- As a USER, I would like to have time delays on payment claiming to allow payment scheduling and to serve as a security delay.
- As a USER, I would like to be able to stake my tokens to create an advanced payment if I don’t have permission to create one.
- As a USER, I would like to be able to make changes to the advanced payment during the process. However, I want it to require collective input after an initial review step.
- As a USER, I would like the payment funding process to be managed by the
available governance processes.
- As a USER, I would like to be able to comment on and engage with other commenters who are participating in a payments governance process.

## Design
- Figma link https://www.figma.com/file/5V8pr7iMwXsT9L3VAZsmUt/Colony-v5?type=design&node-id=3889-462623&mode=design&t=jLgdXbqtPVm8eGRx-4

image

## Implementation
- https://docs.google.com/document/d/1eVNrXV0Q_j24Eqw1mvQWu6DBeggRsbQWknNylBCc-hM/edit#heading=h.1vdyv0p0hlid

## Acceptance Criteria

- **Payment Builder**
For more custom payments involving multiple recipients receiving a number of different tokens at different times.
    - Select a recipient.
    - Select the token to pay, add as many addition tokens as required.
    - Enter an amount for each token type.
    - Enter claim delay if required.
    - Repeat for as many recipients as required.
- **Batch Payments**
    
    Upload a CSV of recipients and tokens and send payments to all in a single Motion.
    
    - Downloadable CSV template file helps guide the user
    - CSV uploaded and processed in the UI.
    - Recipients are viewable and editable after importing.
- **Split Payments**
Divide a pot of funds evenly or unevenly between a set of recipients.
    - Enter in a total balance of tokens and token type to be split
    - Add several recipients to split the payment among.
    - Choose to their split the payment equally, unequally or by reputation.
        - Equally gives each recipient an equal share.
        - Unequally allows any percent to be assigned to a recipient as long as the total adds up to 100%.
    - By reputation sums up the total reputation of each recipient in the
    nominated team and distributes relative to the percentage.
- **Staged Payments**
A payment broken down into separate chunks which may be released in sequence or at achieved milestones. Useful for example to make partial payments upon delivery of agreed milestones.
    - Select a recipient for the staged payment.
    - Be able to enter in the stage milestone description and an amount for the
    milestone.
    - Be able to release any of the staged payments individually and in any order.
- **Streaming Payments**
Continuously stream tokens on a per second basis from multiple teams to a single recipient. Streams can be for a set period of time or indefinite. Useful for things like salaries.
    - Select a recipient to make a streaming payment too.
    - Select a domain source to make the stream from.
    - Enter in a rate amount.
    - Enter in a period.
    - Be able to start the stream.
    - Allow the recipient to claim at any time.

- **General**
    - Should be in alignment with desktop/tablet/mobile design from Designs.
    - Should use existing components and styles where possible for consistency.
    - Error states handled as specified in design specs.
    - All advanced payment action types appear in the activity feed. Following the same sorting and filtering rules with the addition of the following.
        - On creation Advanced payments should appear by default if not using the Reputation decision method to create it, i.e. a 10% staking filter is not required.
        - Sorting
            - Sort by ending soonest. Along with all other action types.
                - This needs to consider the stage that the advanced payment is in and any Reputation decision method processes throughout the payment.
        - Filtering
            - Status (added to existing statuses filters)
                - Review
                - Funding
                - Release
                - Changes
                - Cancel
                - Payment
    - All advanced payment action types should also be displayable grouped with Simple payments in a dedicated “Expenses” table.
        - This is basically just filtering by more then one action type at once.
    - “Open payments” extension refers to the “Staked Expenditures” extension contract.
    - Decision method is included in the creation screen for consistency with other action types.
        - It is used to set the default decision method throughout the following steps in the process.
    - All advanced payment action types appear in the UserHub Transactions and Stakes tabs.
        - For stakes, it should show an accumulation of all stakes of the action.
    
- **Advanced payments process**
    - Users are able to select and create an advanced payment from the action panel.
    - Decision methods based on enabled extensions:
        - Rules for decision methods remain the same as any other actions, with the addition of the following details.
        - Permissions decision method.
            - Without the Open payments extension enabled and without Lazy Consensus extension enabled.
            - The user requires the “Admin” permission or higher (`Administration` on the contracts) to create the payment.
            - If they don’t have permissions, they will receive the disabled action state and an error message.
            - No modal is required when creating.
        - Reputation decision method.
            - When Lazy Consensus extension is enabled.
            - Adds the option to use the Reputation method to create the payment.
            - Goes through the Lazy consensus process before creating the payment.
                - On creation the payment is hidden until 10% is staked.
            - Needs to handle the the team not having reputation and provide validation.
            - No modal is required when creating.
        - Open Payments decision method.
            - When Open Payments extension is enabled.
            - Adds the option to use the “Open payment creation” decision method to create the payment.
                - This decision method is only added to Advanced payment action types.
            - Clicking create opens the Create the payment modal and the user has to stake in order for the payment to be created.
    - Stages
        - Review
            - First stage after creating the payment and staking (if enabled and required), it makes the payment public for others to review.
            - Allows the Owner (creator) to make changes to the payment without requiring collective input.
            - Owner has the option to confirm details, locking the payment details and moving it to the Funding stage.
            - The Owner is also able to cancel the payment at this stage, without requiring collective input through a decision method, and returning their full stake with no penalty.
        - Funding
            - Step to fund the payment with the required funds to make the payment.
            - Opens the Funding modal.
            - Network fee needs to be added on top of the entered amount to ensure the recipient receives their expected funds without the network fee being deducted from their expected payment.
            - Requires “Payer” tiered permissions or higher (`Funding` on the contracts), or using the Reputation decision method to fund the expenditure.
            - On the contract, the payment is in a locked state, changes to the payment at this point will require collective input or the “Admin” tiered permissions or higher (`Arbitration` on the contracts).
            - Funding opens a modal that allows the user see the total amount of what is going to get funded for each token and the network fee. The user also will need to select from one of the decision methods that is available to them.
            - Funding with Permissions decision method:
                - Immediately moves the process to the Release stage.
                - Clicking back on the funding stage pill will show the permission decision method was used.
            - Funding with Reputation decision method:
                - Starts the full lazy consensus process including staking, voting, reveal etc.
                - Completed funding process
                    - Collapses the steps of the process into the collapsed string.
                    - Clicking back on the completed funding stage pill or the collapsed text will show the full process with all pills, clicking on the funding pill again or the collapse text will collapse again.
                - If funding fails from objection, the funding process can be started again.
            - If there is a change to the payment amount that is more then the already funded amount at any time after funding, then the process moves back to the funding stage.
        - Release
            - Opens the Release payment modal with the users possible decision methods.
            - If funded, the owner can release the funds for the payment, starting the claim delay.
            - If any claim delays are 0, then those payments will happen immediately on release and also release the creators stake if present.
            - Alternatively, users can use available decision methods, including “Admin” tiered permissions or higher (`Arbitration` on the contracts), or use the Reputation decision method.
        - Changes
            - Conditional stage that only appears if a change request has been made on the payment after the review stage.
            - Additional details are included in “Changes to advanced payments” below.
        - Cancel
            - Conditional stage that only appears if a cancel request has been made.
            - Additional details are included in “Cancel advanced payments” below.
        - Payment
            - Appears after the funds have been released.
            - Provides a break down of total funds to be paid, paid, and are ready to be paid.
            - Shows a timer that counts down to the next available claim time.
            - Button to “Make payment”
                - Button is disabled if there is nothing to pay out.
                - Button is enabled as long as there is something to pay out. Anyone connected member is able to click and make the payment.
                - First payout will return the creators stake.
            - Button does not appear anymore if there are no more funds to be paid out and there are no claims to be released.
        - Passed
            - A state that represents that no payments are left to be made and no stakes are left to claim.
            - This is not a contract state, but will need to be determined in the app.
    - Flow of process
    
adv-payment-flow
    

- **Create the payment modal**
    - This modal allows the creator to stake the required amount of tokens to create the payment.
    - This modal will only show if the “Open Payments” extension is enabled
    - API - https://github.com/JoinColony/colonyCDapp/pull/860

- **Funding modal**
    - Modal to reviewing the amount to be funded and select the decision method to use.
    - Cancel closes the modal and does nothing. You can also click out to close the modal.
    - Clicking on “Fund payment” will start the funding process, with the process dependent on the decision method chosen.
        - Reputation will start the Staking step.
        - Permissions will immediately fund and move to the release step.
    - Decision methods
        - Options
            - Permissions
                - Visible if the user holds the Payer permissions or Funding custom permission in the team.
            - Reputation
                - Visible to everyone, if the extension is enabled.
        - Selecting a decision method will reveal a callout to explain the next step that will happen based on their selection.

- **Release payment modal**
    - Modal to release the payment.
    - Cancel closes the modal and does nothing. You can also click out to close the modal.
    - Clicking on “Release payment” will release the payment, start claim delays, or start the lazy consensus process dependent on the decision method chosen.
        - Reputation will start the Staking step.
        - Permissions will immediately release the funds and start claim delays.
        - Payment creator will immediately release the funds and start claim delays.
    - Decision methods
        - Options
            - Permissions
                - Visible if the user holds the Payer permissions or Arbitration custom permission in the team.
            - Reputation
                - Visible to everyone, if the extension is enabled.
            - Payment creator
                - Visible to the user who created the payment.
        - Selecting a decision method will reveal a callout to explain the next step that will happen based on their selection.

- **Changes to advanced payments**
    - Payments can me modified freely by the creator of the payment during the review stage.
    - Edit button
        - Visible to the payment creator.
        - Visible to members who hold the Payer level permissions or above or `Arbitration` permission on the contracts.
        - Visible to all members if the Lazy consensus extension is enabled.
        - It is not visible otherwise.
        - Clicking starts the edit mode.
    - Edit mode
        - Updates the read-only fields to be editable for the following fields.
            - Recipients
            - Amounts and tokens
            - Claim delays
            - All other fields should remain read only.
        - Clicking “Propose changes” opens the changes modal.
    - Changes modal
        - Shows the proposed changes, with current and proposed both being visible.
        - Decision methods
            - Options
                - Permissions
                    - Visible if the user holds the Payer permissions or Arbitration custom permission in the team.
                - Reputation
                    - Visible to everyone, if the extension is enabled.
                - Payment creator
                    - This decision method is only visible to the payment creator when in the review stage.
            - Selecting a decision method will reveal a callout to explain the next step that will happen based on their selection.
    - Multiple changes
        - We need to be able to handle multiple changes at once in the UI, listed in the changes step.
        - Clicking each option will reveal the staking card for each.
        - While multiple change proposals can be can be created and staked using the lazy consensus method, as soon as one of them is staked 100%, all others fail and stakes are returned, as only one change process can be active in the process at once.
        - Sorting should be by ending soonest.
    - Change edge case condition:
        - If any variable has previously been voted on, subsequent votes will require more Reputation to be in favour in order for changes to be applied.
            - Reputation is considered on the whole expenditure for each subsequent vote.
            - For example if a change goes to vote for the amount, and 1,000 reputation is used to vote. If another change goes to vote, 1,000+ or more reputation needs to vote for the change to take affect.

- **Cancel advanced payments**
    - Payments can me cancelled freely by the creator of the payment during the review stage and they will get their full stake back.
    - Cancel button
        - Visible to the payment creator.
        - Visible to members who hold the Payer level permissions or above or `Arbitration` permission on the contracts.
        - Visible to all members if the Lazy consensus extension is enabled.
        - It is not visible otherwise.
        - Clicking opens the Cancel modal.
    - Cancel modal
        - During the review stage:
            - Only shows the option to cancel.
            - Can be cancelled by the payment creator or members who hold the Payer level permissions or above or `Arbitration` permission on the contracts.
        - After the review stage:
            - Shows the creator’s stake amount
            - Gives the option to penalize or not penalize the payment creator’s stake.
            - Penalize will mean they lose their full stake and equivalent reputation. Not penalizing will mean they get their stake back.
            - Decision methods
                - Options
                    - Permissions
                        - Visible if the user holds the Payer permissions or Arbitration custom permission in the team.
                    - Reputation
                        - Visible to everyone, if the extension is enabled.
                - Selecting a decision method will reveal a callout to explain the next step that will happen based on their selection.
- **Network fee**
    - Network fee is deducted from the funded expenditure pots
        - To ensure the recipient receives the correct amount, the expenditure should be funded with the expected amount plus the network fee at the time of funding.
        - The UI should still show the expected payment and be able to display the network fee separately where needed.
        - If the network fee changes after funding, and it is higher then the previous fee, then the payment will move back to the funding step. If it is less, it will just return the difference to the colony at the end of the process.
    - Each different token type has to pay their own network fee.
    - Network fee is deducted from the first pot of the same token type, meaning we need to make sure that the network fee is added to the first pot.

### PR DESCRIPTION
**Pull request by jakubcolony**
## Description

This PR contains the first batch of changes made to the advanced payments API to support the new UI.

It consist of the following PRs:
- #1912 
- #1924 
- #1918 
- #1958 
- #1941
- #1983 
- #1984 

Each of the contributing PRs has been tested individually so testing should focus on any regressions that might have been introduced (although none are expected considering this feature is pretty well isolated). 

There are no breaking changes to the database model. A few fields have been removed so we need to confirm this can be deployed safely before pushing to QA and production.

block-ingestor PR: https://github.com/JoinColony/block-ingestor/pull/176

Resolves #1956

### CODE REVIEW

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
Seems like this is mostly still intact! Just a few notes:

#1924 - There was a recent commit to remove decisionMethod field from expenditures - so assume this is no longer required?

#1918 - expenditure balances in the query did not update after finalising (balance on the balance page did update correctly though)

#1958 - unable to test as the create staked expenditure button returns this gasEstimation error in the console: `colony-token-locking-insufficient-deposit`

(I also had to run block-ingestor#176 separately as I was having weird gas estimation issues otherwise, but the commit hash looks correct, so probably just a weird issue with my dev environment)

**Comment by jakubcolony** [comment]
Hey @iamsamgibbs, thanks for your thorough review!

> https://github.com/JoinColony/colonyCDapp/pull/1924 - There was a recent commit to remove decisionMethod field from expenditures - so assume this is no longer required?

That's right, it was supposed to keep track of the user's preferences when it came to funding an expenditure, but the designs have been updated and it's no longer needed.

> https://github.com/JoinColony/colonyCDapp/pull/1958 - unable to test as the create staked expenditure button returns this gasEstimation error in the console: colony-token-locking-insufficient-deposit

Sounds like you don't have enough active tokens - have you activated some before attempting to stake?

> https://github.com/JoinColony/colonyCDapp/pull/1918 - expenditure balances in the query did not update after finalising (balance on the balance page did update correctly though)

Great spot, should be fixed now.

> (I also had to run block-ingestor#176 separately as I was having weird gas estimation issues otherwise, but the commit hash looks correct, so probably just a weird issue with my dev environment)

Not something I could reproduce but give me a shout if you run into those again.

**Comment by iamsamgibbs** [comment]
> Sounds like you don't have enough active tokens - have you activated some before attempting to stake?

🤦‍♂️ You're completely right, I missed that step in the testing instructions. That all works fine, however I wasn't able to cancel and punish an expenditure. Everything else is good though!

**Comment by jakubcolony** [comment]
> That all works fine, however I wasn't able to cancel and punish an expenditure.

Another great spot, the saga was getting stuck due to wrong order of calls, fixed!

File: src/redux/sagas/expenditures/cancelDraftExpenditure.ts [hunk 1]
```diff
@@ -37,40 +53,131 @@ function* cancelDraftExpenditure({
...
+        );
+      }
+
+      const {
+        payload: { hash: txHash },
+      } = yield takeFrom(
+        cancelAndReclaimStake.channel,
+        ActionTypes.TRANSACTION_HASH_RECEIVED,
+      );
+      yield initiateTransaction({ id: cancelAndReclaimStake.id });
```
Comment (iamsamgibbs) [inline src/redux/sagas/expenditures/cancelDraftExpenditure.ts:99]: initiateTransaction needs to be moved to above takeFrom

File: src/redux/sagas/expenditures/cancelStakedExpenditure.ts [hunk 2]
```diff
@@ -64,33 +81,72 @@ function* cancelStakedExpenditure({
...
+    }
+
+    const {
+      payload: { hash: txHash },
+    } = yield takeFrom(
+      cancelStakedExpenditure.channel,
+      ActionTypes.TRANSACTION_HASH_RECEIVED,
+    );
+
+    yield initiateTransaction({ id: cancelStakedExpenditure.id });
```
Comment (iamsamgibbs) [inline src/redux/sagas/expenditures/cancelStakedExpenditure.ts:116]: And here too

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
Left a comment which is preventing 'Cancel and punish' from working (works fine when reordered), otherwise all good!

**Comment by jakubcolony** [comment]
@iamsamgibbs fixed

**Review by iamsamgibbs — APPROVED** [review]
All good now! 🎉

**Review by ArmandoGraterol — DISMISSED** [review]
LGTM 👌

**Pull request merged**